projects/continuous_learning/dend_kwinners.py

```diff
@@ -0,0 +1,195 @@
+
+
+def update_boost_strength(m):
+    """Function used to update KWinner modules boost strength. This is typically done
+    during training at the beginning of each epoch.
+
+    Call using :meth:`torch.nn.Module.apply` after each epoch if required
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, DKWinnersBase):
+        m.update_boost_strength()
+
+
+class DKWinnersBase(nn.Module, metaclass=abc.ABCMeta):
+
+    def __init__(
+        self,
+        out_dim=10,
+        dpc=3,
+        k_inference_factor=1.0,
+        boost_strength=1.0,
+        boost_strength_factor=1.0,
+        duty_cycle_period=1000,
+        new_epoch=False,
+        per_epoch=False,
+    ):
+        super(DKWinnersBase, self).__init__()
+        assert boost_strength >= 0.0
+        assert 0.0 <= boost_strength_factor <= 1.0
+
+        self.k_inference_factor = k_inference_factor
+        self.learning_iterations = 0
+        self.n = 0
+        self.k = 0
+        self.k_inference = 0
+        self.dpc = dpc,
+
+        # Boosting related parameters
+        self.register_buffer("boost_strength", torch.tensor(boost_strength,
+                                                            dtype=torch.float))
+        self.boost_strength_factor = boost_strength_factor
+        self.duty_cycle_period = duty_cycle_period
+        self.new_epoch = new_epoch
+        self.per_epoch = per_epoch
+
+    def extra_repr(self):
+        return (
+            "n={0}, dpc={1}, boost_strength={2}, boost_strength_factor={3}, "
+            "k_inference_factor={4}, duty_cycle_period={5}".format(
+                self.n, self.dpc, self.boost_strength,
+                self.boost_strength_factor, self.k_inference_factor,
+                self.duty_cycle_period
+            )
+        )
+
+    @abc.abstractmethod
+    def update_duty_cycle(self, x):
+        r"""Updates our duty cycle estimates with the new value. Duty cycles are
+        updated according to the following formula:
+
+        .. math::
+            dutyCycle = \frac{dutyCycle \times \left( period - batchSize \right)
+                                + newValue}{period}
+
+        :param x:
+          Current activity of each unit
+        """
+        raise NotImplementedError
+
+    def update_boost_strength(self):
+        """Update boost strength by multiplying by the boost strength factor.
+        This is typically done during training at the beginning of each epoch.
+        """
+        self.boost_strength *= self.boost_strength_factor
+
+    def entropy(self):
+        """Returns the current total entropy of this layer."""
+        _, entropy = binary_entropy(self.duty_cycle)
+        return entropy
+
+    def max_entropy(self):
+        """Returns the maximum total entropy we can expect from this layer."""
+        return max_entropy(self.n, int(self.n * self.percent_on))
+
+
+class DKWinners(DKWinnersBase):
```

Comment:
There's no docstring to explain what it's supposed to do.
DKWinners as a name is completely incomprehensible - use a better name.